src/PerfCounters.cc

```diff
@@ -1118,4 +1125,166 @@ Ticks PerfCounters::read_ticks(Task* t, Error* error) {
   return ret;
 }
 
+
+class BpfAccelerator {
+public:
+  static std::shared_ptr<BpfAccelerator> get_or_create();
+
+  ScopedFd create_counter(pid_t tid);
+  void match_regs_and_open_counter(Registers regs, ScopedFd& counter);
+  uint64_t skips() const {
+    return *bpf_skips;
+  }
+
+  // Can't be private because of make_shared.
+  BpfAccelerator(ScopedFd&& bpf_prog_fd, user_regs_struct* bpf_regs, uint64_t* bpf_skips)
+    : bpf_prog_fd(std::move(bpf_prog_fd)), bpf_regs(bpf_regs), bpf_skips(bpf_skips)
+  {}
+
+private:
+  static std::shared_ptr<BpfAccelerator> singleton;
+
+  struct perf_event_attr attr;
+  ScopedFd bpf_prog_fd;
+  user_regs_struct* bpf_regs;
+  uint64_t* bpf_skips;
+};
+
+std::shared_ptr<BpfAccelerator> BpfAccelerator::singleton;
+
+#ifdef BPF
+/* static */ std::shared_ptr<BpfAccelerator> BpfAccelerator::get_or_create() {
+  static int initialized;
+  if (BpfAccelerator::singleton) {
+    return BpfAccelerator::singleton;
+  }
+
+  if (!initialized) {
+    initialized = -1;
+
+    libbpf_set_strict_mode(LIBBPF_STRICT_DIRECT_ERRS);
+    string path = resource_path() + "share/rr/async_event_filter.o";
+    struct bpf_object* obj = bpf_object__open(path.c_str());
+    if ((intptr_t)obj <= 0) {
+      return nullptr;
```

Comment:
Why not do a `FATAL()` here with an appropriate message? Same for the other errors below? Seems to me we don't want mysterious performance drops.

Reply:
*Maybe* it's OK to silently fail if we don't have eBPF capability, but the other errors should probably be fatal.

Reply:
I think bpf_object__load is the one that would fail since that's the one that loads it into the kernel.